src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -491,109 +499,110 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
 
                     var state = ReadState();
 
-                    // if we already have a compilation, we must be already done!  This can happen if two
-                    // threads were waiting to build, and we came in after the other succeeded.
-                    if (state is FinalState finalState)
-                        return finalState;
-
-                    var compilationWithoutGeneratedDocuments = state.CompilationWithoutGeneratedDocuments;
-                    if (compilationWithoutGeneratedDocuments == null)
+                    return state switch
                     {
+                        // if we already have a compilation, we must be already done!  This can happen if two
+                        // threads were waiting to build, and we came in after the other succeeded.
+                        FinalState finalState
+                            => finalState,
+
                         // We've got nothing.  Build it from scratch :(
-                        return await BuildFinalStateFromScratchAsync(
-                            compilationState,
-                            state.GeneratorInfo,
-                            cancellationToken).ConfigureAwait(false);
-                    }
+                        NoCompilationState
+                            => await BuildFinalStateFromScratchAsync(
+                                compilationState,
+                                state.GeneratorInfo,
+                                cancellationToken).ConfigureAwait(false),
 
-                    if (state is AllSyntaxTreesParsedState)
-                    {
                         // We have a declaration compilation, use it to reconstruct the final compilation
-                        return await FinalizeCompilationAsync(
-                            compilationState,
-                            compilationWithoutGeneratedDocuments,
-                            state.GeneratorInfo,
-                            compilationWithStaleGeneratedTrees: null,
-                            cancellationToken).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        // We must have an in progress compilation. Build off of that.
-                        return await BuildFinalStateFromInProgressStateAsync(
-                            compilationState, (InProgressState)state, compilationWithoutGeneratedDocuments, cancellationToken).ConfigureAwait(false);
```

Comment:
this was passing in compilationWithoutGeneratedDocuments which was retrieved off of `state` anyways.  so we can just pass in `state` anyways.